File: routes/fireflyVideoRecord.js
/**
 * Web2Media Service — Firefly Video Record Route Handlers
 */

const express = require('express');
const fs = require('fs');
const router = express.Router();

const { validateFireflyVideoRecordRequest } = require('../middleware/validator');
const { renderFireflyVideo, getActiveFireflyVideoRecordCount } = require('../services/renderer');
const { convertWithOptionalAudio, getMimeType } = require('../services/converter');
const { uploadFileToR2 } = require('../services/r2Uploader');
const { getRuntimeR2Config } = require('../services/runtimeConfig');
const { BG_PRESETS, COLOR_PRESETS, DIRECTIONS, GLOW_LEVELS } = require('../config');

/**
 * POST /api/firefly-video-record
 * Create a firefly animation video with given configuration.
 * 
 * Returns the video file as a download.
 */
router.post('/firefly-video-record', validateFireflyVideoRecordRequest, async (req, res) => {
  const params = req.body;
  const hasAudio = params.audioUrls.length > 0;
  let r2Config = null;
  let webmPath = null;
  let outputPath = null;

  try {
    if (hasAudio && params.format === 'gif') {
      return res.status(400).json({
        success: false,
        error: 'Không thể ghép audio với GIF. Vui lòng chọn format "mp4" hoặc "webm".',
      });
    }

    const startTime = Date.now();

    console.log(`\n[API] ═══ New firefly video record request ═══`);
    console.log(`[API] Config: ${params.count} fireflies, ${params.size} size, ${params.speed}x speed`);
    console.log(`[API] Color: ${params.colorMode === 'preset' ? COLOR_PRESETS[params.colorIndex]?.name : params.customColor}`);
    console.log(`[API] Background: ${params.bgUrl ? 'URL → ' + params.bgUrl : 'Preset #' + params.bgIndex}`);
    console.log(`[API] Video output: ${params.duration}s, ${params.width}x${params.height}, ${params.fps}fps, ${params.format}`);
    if (hasAudio) {
      console.log(`[API] Audio URLs: ${params.audioUrls.length}`);
    }
    console.log(`[API] Active firefly video records: ${getActiveFireflyVideoRecordCount()}`);

    console.log(`[API] Loading R2 runtime config...`);
    try {
      r2Config = await getRuntimeR2Config();
      console.log(`[API] R2 upload: ${r2Config.bucket}`);
    } catch (runtimeConfigErr) {
      runtimeConfigErr.statusCode = 502;
      throw runtimeConfigErr;
    }

    // Step 1: Render video with Puppeteer
    console.log(`[API] Step 1/2: Rendering video...`);
    webmPath = await renderFireflyVideo(params);

    // Step 2: Convert to requested format, optionally with audio
    console.log(`[API] Step 2/2: ${hasAudio ? 'Converting and merging audio' : `Converting to ${params.format}`}...`);
    outputPath = await convertWithOptionalAudio(webmPath, params.format, {
      bitrate: params.bitrate,
      fps: params.fps,
      width: params.width,
      audioUrls: params.audioUrls,
    });
    webmPath = null;

    const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
    const fileSize = fs.statSync(outputPath).size;
    console.log(`[API] ✓ Done in ${elapsed}s — ${(fileSize / 1024 / 1024).toFixed(2)} MB`);

    const filename = `${params.filename}.${params.format}`;
    const mimeType = getMimeType(params.format);

    console.log(`[API] Step 3/3: Loading R2 runtime config and uploading...`);
    let uploadResult;
    try {
      uploadResult = await uploadFileToR2(outputPath, r2Config, {
        filename,
        contentType: mimeType,
      });
    } catch (uploadErr) {
      uploadErr.statusCode = 502;
      throw uploadErr;
    }

    try { fs.unlinkSync(outputPath); } catch (e) { /* ignore */ }
    outputPath = null;

    return res.json({
      success: true,
      url: uploadResult.url,
      data: {
        url: uploadResult.url,
        filename,
        format: params.format,
        mimeType,
        size: fileSize,
        duration: params.duration,
        hasAudio,
        r2: uploadResult,
      },
    });

  } catch (err) {
    const errMsg = err?.message || String(err) || 'Lỗi không xác định khi tạo video';
    console.error(`[API] ✗ Error:`, errMsg);

    // Cleanup on error
    if (outputPath) {
      try { fs.unlinkSync(outputPath); } catch (e) { /* ignore */ }
    }
    if (webmPath && webmPath !== outputPath) {
      try { fs.unlinkSync(webmPath); } catch (e) { /* ignore */ }
    }

    const statusCode = err.statusCode || (errMsg.includes('giới hạn') ? 429 : 500);
    res.status(statusCode).json({
      success: false,
      error: errMsg,
    });
  }
});

/**
 * GET /api/presets
 * Return available presets for backgrounds, colors, directions, and glow levels.
 */
router.get('/presets', (req, res) => {
  res.json({
    success: true,
    data: {
      backgrounds: BG_PRESETS.map(({ index, label }) => ({ index, label })),
      colors: COLOR_PRESETS.map(({ index, name, hex }) => ({ index, name, hex })),
      directions: DIRECTIONS,
      glowLevels: GLOW_LEVELS,
      fpsOptions: [24, 30, 60],
      formatOptions: ['webm', 'mp4', 'gif'],
    },
  });
});

/**
 * GET /api/health
 * Health check endpoint.
 */
router.get('/health', (req, res) => {
  res.json({
    success: true,
    status: 'ok',
    version: '1.0.0',
    activeFireflyVideoRecords: getActiveFireflyVideoRecordCount(),
    timestamp: new Date().toISOString(),
  });
});

module.exports = router;

File: app/api/firefly_video_record.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pydantic import ValidationError

from app.core.config import BG_PRESETS, COLOR_PRESETS, DIRECTIONS, GLOW_LEVELS
from app.schemas.firefly import normalize_firefly_request, validation_error_response
from app.services.jobs import JobStatus, job_queue
from app.services.renderer import get_active_firefly_video_record_count


router = APIRouter()


async def read_json_body(request: Request) -> dict[str, Any]:
    try:
        body = await request.json()
    except Exception:
        return {}
    return body if isinstance(body, dict) else {}


@router.post("/firefly-video-record")
async def create_firefly_video_record(request: Request) -> JSONResponse:
    payload = await read_json_body(request)
    try:
        params = normalize_firefly_request(payload)
    except ValidationError as err:
        return JSONResponse(status_code=400, content=validation_error_response(err))

    has_audio = bool(params["audioUrls"])
    if has_audio and params["format"] == "gif":
        return JSONResponse(
            status_code=400,
            content={
                "success": False,
                "error": 'Khong the ghep audio voi GIF. Vui long chon format "mp4" hoac "webm".',
            },
        )

    job = await job_queue.create_job(params)
    status_url = f"/api/firefly-video-record/status/{job['jobId']}"
    print(f"[API] Firefly video job queued: {job['jobId']}")

    return JSONResponse(
        status_code=202,
        content={
            "success": True,
            "jobId": job["jobId"],
            "status": job["status"],
            "statusUrl": status_url,
            "data": {**job, "statusUrl": status_url},
        },
    )


async def send_firefly_video_job_status(job_id: str) -> JSONResponse:
    job = await job_queue.get_job(job_id)
    if not job:
        return JSONResponse(
            status_code=404,
            content={
                "success": False,
                "error": "Khong tim thay job",
                "jobId": job_id,
            },
        )

    return JSONResponse(
        content={
            "success": job["status"] != JobStatus.FAILED,
            **job,
        }
    )


@router.get("/firefly-video-record/status/{job_id}")
async def get_firefly_video_job_status(job_id: str) -> JSONResponse:
    return await send_firefly_video_job_status(job_id)


@router.get("/firefly-video-record/{job_id}/status")
async def get_firefly_video_job_status_alias(job_id: str) -> JSONResponse:
    return await send_firefly_video_job_status(job_id)


@router.get("/firefly-video-record/{job_id}")
async def get_firefly_video_job_status_short(job_id: str) -> JSONResponse:
    return await send_firefly_video_job_status(job_id)


@router.get("/presets")
async def get_presets() -> dict[str, Any]:
    return {
        "success": True,
        "data": {
            "backgrounds": [{"index": item["index"], "label": item["label"]} for item in BG_PRESETS],
            "colors": [{"index": item["index"], "name": item["name"], "hex": item["hex"]} for item in COLOR_PRESETS],
            "directions": DIRECTIONS,
            "glowLevels": GLOW_LEVELS,
            "fpsOptions": [24, 30, 60],
            "formatOptions": ["webm", "mp4", "gif"],
        },
    }


@router.get("/health")
async def health() -> dict[str, Any]:
    return {
        "success": True,
        "status": "ok",
        "version": "1.0.0",
        "activeFireflyVideoRecords": get_active_firefly_video_record_count(),
        "fireflyVideoJobs": await job_queue.get_stats(),
    }
